cpp/include/cudf/ast/detail/operators.hpp

```diff
@@ -69,159 +69,111 @@ constexpr bool is_valid_unary_op = cuda::std::is_invocable_v<Op, T>;
  * @param args Forwarded arguments to `operator()` of `f`.
  */
 template <typename F, typename... Ts>
-CUDF_HOST_DEVICE inline constexpr void ast_operator_dispatcher(ast_operator op, F&& f, Ts&&... args)
+CUDF_HOST_DEVICE inline constexpr decltype(auto) ast_operator_dispatcher(ast_operator op,
+                                                                         F&& f,
+                                                                         Ts&&... args)
 {
   switch (op) {
     case ast_operator::ADD:
-      f.template operator()<ast_operator::ADD>(std::forward<Ts>(args)...);
-      break;
+      return f.template operator()<ast_operator::ADD>(std::forward<Ts>(args)...);
```

Comment:
could we also benefit from using indices into function pointers instead of the switch statements for dynamic operator dispatch?

Reply:
How much did the performance change? Can you share your micro-benchmark and the results?

One of the biggest challenges we face with the AST is compile time and kernel size -- if the function-pointer dispatch helps with either of those, it could still be worth evaluating.